external_executors/dnanexus_wes/dx-wes-lambda/dx_wes.py

```diff
@@ -0,0 +1,510 @@
+    # PATH and PYTHONPATH so we can continue using the dx-toolkit dependencies.
+    stdout = run_and_capture_log(
+        dxwdl_cmd,
+        cwd="/tmp",
+        env={"DX_SECURITY_CONTEXT": auth_header,
+             "DX_PROJECT_CONTEXT_ID": event["dx_project_id"],
+             "PYTHONPATH": ':'.join([os.environ.get("PYTHONPATH", ""),
+                                     os.path.dirname(os.path.dirname(dx_exe_path))]),
+             "PATH": ':'.join([os.environ["PATH"], os.path.dirname(dx_exe_path)])}
+    )
+
+    dx_workflow_id = stdout.strip()
+
+    # This file should have been created by dxWDL.
+    dx_inputs = json.load(open("/tmp/dx_inputs.dx.json"))
+
+    # Finally, actually run the newly created DNAnexus workflow and return its
+    # analysis ID. "Workflow" in WES means "analysis" on DNAnexus.
+    dx_workflow = dxpy.DXWorkflow(dx_workflow_id)
+    dx_analysis = dx_workflow.run(dx_inputs, project=event["dx_project_id"])
+
+    # Finally, we need to record the ID that DNAnexus returned along with the
+    # ID that was generated immediately when the WES request was made.
+    record_dx_analysis_id(event["wes_workflow_id"], dx_analysis.id)
+    record_dxwdl_log(event["wes_workflow_id"], "Hey it looks like everything worked.")
+
+
+def dnanexus_workflows_post(event, context):
+    """Handle a WES workflows POST request and turn it into a DNAnexus
+    /{workflow-id}/run request.
+
+    Mostly delegates everything to run_dxwdl so this function can return right
+    away.
+    """
+
+    # Create a workflow ID that we can return right away. It will likely take
+    # longer than 29 seconds to complete data localization and workflow
+    # creation, so we have to delegate that to another function.
+    wes_workflow_id = str(uuid.uuid4())
+
+    # Record the workflow id we just generated in our table so we can look it
+    # up later. For now just record an empty DX id since nothing's been created
+    # on the DNAnexus platform yet.
+    record_dx_analysis_id(wes_workflow_id, "None")
+
+    # Now we pass all the work to another lambda that's going to do the actual
+    # work. First prepare the inputs:
+    dxwdl_payload = {
+        "workflow_descriptor": event["body"]["workflow_descriptor"],
+        "workflow_params": event["body"]["workflow_params"],
+        "workflow_dependencies": event["body"]["workflow_dependencies"],
+        "auth_header": event["headers"]["Authorization"],
+        "dx_project_id": event["body"]["key_values"]["dx-project"],
+        "wes_workflow_id": wes_workflow_id
+    }
+
+    lambda_client = boto3.client('lambda')
+
+    # Note the "Event" InvocationType means that this is executed asychronously
+    lambda_client.invoke(
+        FunctionName=os.environ["AWS_DXWDL_LAMBDA_FN"],
+        InvocationType="Event",
+        Payload=json.dumps(dxwdl_payload))
+
+    return {"workflow_id": wes_workflow_id}
+
+
+def dnanexus_workflows_get_status(event, context):
```

Comment:
I'm a bit confused about how this request is supported because of how it seems to rely on the state of dxpy. This suggests to me you're hitting a single instance with these requests. 

So, stepping back, you have a lambda that starts up an instance for dxWDL when called, and it starts running a workflow. Is the backend for the executed tasks not local? 

Otherwise, lets say you have multiple workflows running.. are they all executing on the same instance?

Reply:
I think all the state should be stored in the dynamodb table, and to the extent there is any state persistence across function executions, it's an unfortunate side effect of lambda container reuse.

The way it works is that a lambda runs dxWDL. dxWDL creates a "workflow" object on dnanexus. It also starts the execution of that workflow, creating what dnanexus calls an "analysis". One the workflow execution is started, the dxWDL function terminates while the workflow is running in dnanexus.

However, we recorded the dnanexus id of the analysis, so later we can use the dnanexus API to ask questions about it. In this function, we look up that id, and make [this call](https://wiki.dnanexus.com/API-Specification-v1.0.0/Workflows-and-Analyses#API-method%3A-%2Fanalysis-xxxx%2Fdescribe) to find out information about it. `dxpy` is just the python bindings to making that API request; it's not storing any state across executions.